renku/core/management/datasets.py

```diff
@@ -126,42 +128,23 @@ def renku_pointers_path(self):
         path.mkdir(exist_ok=True)
         return path
 
-    @property
-    def datasets_provenance(self):
-        """Return dataset provenance if available."""
-        if not self.has_datasets_provenance():
-            return
-        if not self._datasets_provenance:
-            self._datasets_provenance = DatasetProvenance.from_json(self.datasets_provenance_path)
-
-        return self._datasets_provenance
-
-    def update_datasets_provenance(self, dataset, remove=False):
+    @inject.autoparams()
+    def update_datasets_provenance(
```

Comment:
Would it be cleaner to have two methods, `add_or_update_dataset_provenance` and `remove_dataset_provenance`?

It would also be cleaner to just call the methods on `DatasetProvenance` directly and inject that where needed, since this method in LocalClient doesn't really contain any additional logic (especially since `commit_database` isn't set to `True` anywhere). In the future I'd like to rely on `LocalClient` less and have the dedicated classes do more of the heavy lifting.
If we inject `DatasetProvenance` using the `binder.bind_to_constructor(Class, lambda: Class(*args))` way of injection, it would only initialise `DatasetProvenance` if code that uses it (well, requests it as an injection) is called and then it would be a singleton, so we could just inject this, `ProvenanceGraph` and `DependencyGraph` this way in the `with_database` CommandBuilder without performance overhead, and it'd save performance since we wouldn't have to do `from_database` more than once.